### PR TITLE
Introduce login and user management functionality

### PR DESCRIPTION
**Pull request by pariveda-btang**
Changes 
- Introduce user authentication and authorization functionality using Passport.js 
- APIs require JWT passed as Bearer tokens for secured endpoints.
- Introduce roles (User, Admin, Superadmin) associated to user management features in the app.
- Introduce use of enums for roles and frequently used error messages.
- Introduce safeguards to help prevent SQL injection (escaping user inputs, parameterized queries, basic type checking)
- Remove public facing sign up page and replace with modal

Details 
- JWT should be passed as a Bearer token in header for all secured endpoints. 
- Admin/Superadmin have the ability to reset passwords for all users, regular users can only reset their own password. 
- Superadmin and admins have the same privileges. The only exception is that superadmins cannot have their roles revoked and their accounts cannot be removed. 
- Users are expected to contact an admin if they have forgotten their password. 
- Passwords must be at least 8 characters long and contain 1 lowercase letter, 1 uppercase letter, 1 digit, and 1 special character. 
- No JWT refresh has been configured. User will need to log in again after their token expires. Expiry is current set to 20 minutes. 
- Sign up page has been removed and users must contact an admin to register for an account. Since the app will be deployed publicly, we do not user registration to be open. 
- All UI pages containing data require authentication. Users will be redirected to the login page if not authenticated. 

Notes 
- Additional enhancements have been introduced to the README files - these changes do not significantly change the application's functionality but will improve code architecture. 
- Safeguards have been introduced however there are no guarantees that the app is 100% safe against SQL injection

Screenshots
Login page
image
User Management Features (available only to admin)
image

### CODE REVIEW

**Comment by pariveda-btang** [comment]
A small issue that was just found but not breaking. 

When we’re updating or deleting any records (laptops, dropdowns, etc.), there is no check to see if the record exists before running the query. Within SQL, if we execute that query and something doesn’t exist, the result is still a success since it’s technically still correct if nothing matches the identifier. Our API, however, is returning a 200 status stating the operation succeeded, even though nothing was updated/deleted in the DB. 

I don’t expect this will be an issue if users are going through the UI, as that ensures the selected values are valid but if someone decides to mess with the API directly, it’ll execute the query, make no changes to the DB, and return a status that may misinform the user. 

I’ll document this and add it as a Future Enhancement, as it currently does not break anything.

File: backend/index.js [hunk 1]
```diff
@@ -108,70 +143,281 @@ function createLaptopBody(req){
...
+    if (results.length === 0) {
+      return res.status(404).send("Laptop with serial number", id, "not found.");
+    }
+    return res.send(results[0]);    
+  });
+});
+
+// Add a new laptop 
+// TODO: Consider adding type checks once laptop class is better defined
+app.post("/add", authJwt, (req, res, next) => {
```
Comment (alex-seo-pariveda) [inline backend/index.js:335]: Are these parameters use or is it handled in the middleware now?

File: db/create_admin.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+if (typeof username !== "string" || typeof password !== "string" || typeof firstName !== "string" || typeof lastName !== "string") {
+  console.log("Invalid type in one or more parameters.");
+  return;
+}
+if (!ALPHANUMERIC_REGEX.test(firstName) || !ALPHANUMERIC_REGEX.test(lastName)) { 
+  console.log("Invalid characters in firstname and/or lastname.");
+  return;
+}
+if (!PASSWORD_REGEX.test(password)) { 
+  console.log("Password does not meet complexity requirements.");
```
Comment (alex-seo-pariveda) [inline db/create_admin.js:43]: Can we update the error message to list out the complexity requirements? Doesn't have to be dynamic, just list them all

File: backend/enums/errors.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+const errors = {
+    INVALID_PARAMETERS: 'Invalid parameters in request.',
+    ALPHANUMERIC_ONLY: 'Parameters must be alphanumeric only.',
+    PASSWORD_COMPLEXITY: 'Password does not meet the complexity requirements.',
```
Comment (alex-seo-pariveda) [inline backend/enums/errors.js:4]: Lets list out the password complexity requirements in this error message.

File: ui/src/enums/errors.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const errors = { 
+    EMPTY_FIELD: 'Please fill out all fields before submitting.',
+    INVALID_LOGIN: 'Incorrect username and/or password.',
+    USER_EXISTS: 'User already exists.',
+    PASSWORD_COMPLEXITY: 'Password does not meet the complexity requirements.',
```
Comment (alex-seo-pariveda) [inline ui/src/enums/errors.js:5]: Same here

**Review by alex-seo-pariveda — COMMENTED** [review]
Thank you for updating the readmes and cleaning up a lot of the repo!!

File: backend/index.js (hunk 1 above)
Comment (pariveda-btang) [reply backend/index.js:335]: Any references to authJwt or authLocal indicate that the route requires authentication (either through JWT or local username + password). This isn't used as a parameter necessarily but acts as a flag for the middleware to intercept.

**Comment by pariveda-btang** [comment]
Merging, as per offline discussion.

**Pull request merged**